### ISSUE TITLE
New Variant is automatically generated (without user input) when navigating away from an assessment after attempts have been exhausted or after question is correct

### ISSUE THREAD

**Issue by firasm** [PrairieLearn/PrairieLearn#6662]
Full credit to @ffund for discovering this behaviour and pointing us to it.

**tl;dr: After your attempts are exhausted, navigating away from the problem and coming back to it automatically generates a new variant.** 

If I have a question that has 5 attempts, and I end up at the "Last Attempt", the Variant ID is added to my URL (this is probably not desirable): https://ca.prairielearn.com/pl/course_instance/2439/instance_question/39146740/?variant_id=7934934

![Screenshot 2022-11-28 at 1 56 39 PM](https://user-images.githubusercontent.com/2507459/204389425-8d4454ae-cc64-4707-b613-65566dfa4ee7.png)

If I navigate away from the page and come back to it, the Variant ID is removed from the address:

![Screenshot 2022-11-28 at 1 57 17 PM](https://user-images.githubusercontent.com/2507459/204389510-71e070fb-f480-4299-a06e-9f0cfd5db4c1.png)

Exhausting my attempts brings up the "Try a new Variant" button **and** the variant ID is back in the URL (reloading the page keeps variant ID in the URL):

![Screenshot 2022-11-28 at 1 58 11 PM](https://user-images.githubusercontent.com/2507459/204389664-9ee0c9a7-542d-4117-820e-edc96d0604b0.png).

Navigating away from the page, and coming back to it removes my old answers and basically generates a new issue:

![Screenshot 2022-11-28 at 1 59 16 PM](https://user-images.githubusercontent.com/2507459/204389866-cce0243b-5632-4fe9-b518-7a4fe470ff26.png)

**Comment by nwalters512** [comment]
We mark variants as closed if they belong to a homework and if all attempts have been exhausted:

https://github.com/PrairieLearn/PrairieLearn/blob/3b727660b102768a95dc2ebc1fca876e457dfd38/sprocs/variants_update_after_grading.sql#L32-L36

Then, when we go to render an instance question for a homework, we set a flag that mandates that the variant be open:

https://github.com/PrairieLearn/PrairieLearn/blob/3b727660b102768a95dc2ebc1fca876e457dfd38/lib/question.js#L1569

AFAICT, this is undesirable behavior, and we should completely remove `require_open` from `ensureVariants`. @mwest1066 does that sound reasonable to you?

**Comment by mwest1066** [comment]
It's intended behavior that navigating away from a closed (completed) homework question and returning to that question generates a new variant. I'm open to discussion about whether this _should_ be the intended behavior, of course.

The intended usage pattern with HW questions is to encourage repeated attempts on different variants. For this reason, returning to a completed question gives you a new variant, rather than showing you an old variant that you already attempted. It's not clear to me what the value is in seeing an old previously-attempted variant by default.

I do think that it's a good idea to provide access to previously-attempted variants in case students want to return to them and I like @ffund's PR for this: #6105

**Comment by nwalters512** [comment]
I can understand the rationale behind it, but it does feel like unexpected behavior, especially since there's the "New variant" button that can be clicked (and that's presumably the only action that *can* be taken on a closed variant).

As for the value: I can see someone wanting to e.g. cross-reference a failed attempt with another question on the homework. More broadly, I think that the choice to move on to a new variant should the choice of the student.

I agree that something like #6105 will probably make most of this discussion moot since it'll be possible to manually click around to old variants, but even then I'd like to change this behavior.

**Comment by firasm** [comment]
hmm interesting - I just tried it, and it seems a new variant is generated upon navigating away from the page **even when the attempts aren't exhausted** (I'll update the issue title).

I'd like to advocate for students seeing their submissions until they explicitly click "New Variant" for the reasons that Nathan pointed to above. Additionally, in case students make a calculation error on the last part, they need to do all parts of a multi-part question again, which just adds to their frustration. 

I agree the functionality to see info from variants makes a lot of this moot, but I agree, I think it should be the student's choice *when* to generate a new variant.

**Comment by firasm** [comment]
P.S. Devil's advocate argument (against changing current behaviour): some instructors may be relying on this behaviour to avoid having answers to old questions readily available to students on-demand. 

Two reasons why this may be bad:

1. They may need to add more questions to their bank to keep it "as random" (Though, once released, I think questions can be screenshot'd and are in the wild)
1. students don't actually end up trying the problem again and instead just do a quick calculation on their calculator or in their head and convince themselves they know how to do it. There's something nice about actually *doing it again*.

There are probably good arguments for and against...

**Comment by mwest1066** [comment]
I think we all agree that #6105 is the right way to go and will make this discussion much less important one way or the other. That being said:

> As for the value: I can see someone wanting to e.g. cross-reference a failed attempt with another question on the homework. More broadly, I think that the choice to move on to a new variant should the choice of the student.

I do see the strength of this argument and I'm generally in favor of minimizing "magic" behavior in UIs. However, in this case, I'm worried that requiring a "new variant" click even on old homework questions will encourage poor study behavior. At the moment, when a student comes back to a homework to study for an exam, clicking through the questions shows them new variants that are just waiting to be solved. If they instead see their old variants with answers already in them, it's far too easy to say "oh yes, I could clearly do this question, no need to do it again". @firasm captures this point exactly:

> 2. students don't actually end up trying the problem again and instead just do a quick calculation on their calculator or in their head and convince themselves they know how to do it. There's something nice about actually _doing it again_.

This is the key point for me. We really really want students to "do it again". One of the key strengths of PL is getting students to _actually do_ lots of problems. At all costs we want to avoid students studying by looking at solved problems, which we know is close to useless in terms of learning. It's _doing_ that leads to learning.

In class data we see lots of repeated attempts at homework questions after the homework due date but before the exam, when students are repeating problems to make sure they can actually do them. Having a well-randomized question constantly giving you new variants every time you click past the question is a great way to induce you to make sure you can also do _this_ new variant.

To reiterate: I'm not necessarily saying that we shouldn't make this change, but I do think that we'd be better off getting #6105 in and seeing whether that takes care of problem while keeping the benefits of the current UI.

**Comment by nwalters512** [comment]
@firasm @ffund given that #6105 is now (finally) merged and deployed, have your thoughts on this issue changed at all?

**Comment by mwest1066** [comment]
Closing as resolved by the new UI introduced in #6105. If anyone still feels the original issue is still a problem, please comment below and we can continue discussion.
